### PR TITLE
[relations]: Reorder when creating new entity

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

- Adds the reordering logic when creating a new entity. Before we only had it for updating it (it was enough for the CM to work).
- Added some modifications proposed in the [last PR](https://github.com/strapi/strapi/pull/14709)

### Why is it needed?

We are going to expose this logic to the content API, so this logic is also necessary when creating an entity.

### How to test it?

See API tests.

### CODE REVIEW

File: packages/core/strapi/tests/api/relations.test.api.js [hunk 1]
```diff
@@ -132,6 +133,86 @@ const updateEntry = async (pluralName, id, data, populate) => {
   return body.data;
 };
 
+const createShop = async ({
```
Comment (Marc-Roig) [inline packages/core/strapi/tests/api/relations.test.api.js:136]: Imo these helpers make the tests more readable and short. 
If we like this structure, I would like to change tests from this file to use it.

File: packages/core/strapi/tests/api/relations.test.api.js [hunk 2]
```diff
@@ -687,4 +768,60 @@ describe('Relations', () => {
...
   });
+
+  // TODO: Use createShop and updateShop in other tests
+  describe.only('Reorder entity relations', () => {
+    test('Connect new relation at the start', async () => {
+      // Add two products at the start
+      const createdShop = await createShop({
+        anyToManyRel: [
+          { id: id1, position: { start: true } },
+          { id: id2, position: { start: true } },
```
Comment (Marc-Roig) [inline packages/core/strapi/tests/api/relations.test.api.js:779]: add another test case to test it is reordering when creating a new entity, and change start:true to before: id1

File: packages/core/strapi/tests/api/relations.test.api.js [hunk 3]
```diff
@@ -687,4 +768,60 @@ describe('Relations', () => {
       });
     });
   });
+
+  // TODO: Use createShop and updateShop in other tests
+  describe.only('Reorder entity relations', () => {
```
Comment (Marc-Roig) [inline packages/core/strapi/tests/api/relations.test.api.js:773]: remove .only

File: packages/core/database/lib/entity-manager/index.js [hunk 4]
```diff
@@ -579,11 +579,23 @@ const createEntityManager = (db) => {
           });
 
           // add order value
-          if (hasOrderColumn(attribute)) {
-            insert.forEach((rel, idx) => {
-              rel[orderColumnName] = idx + 1;
+          if (cleanRelationData.set && hasOrderColumn(attribute)) {
```
[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/database/lib/entity-manager/index.js:582]: I will create it, it surely needs to be splited into more functions / files.

File: packages/core/database/lib/entity-manager/relations-orderer.js [hunk 5]
```diff
@@ -33,14 +33,14 @@ const _ = require('lodash/fp');
  * @param {number} [maxOrder=0] - Used to calculate the order of relations placed at the end
```
Comment (jhoward1994) [inline packages/core/database/lib/entity-manager/relations-orderer.js:33]: I think this JSDoc param should be removed

File: packages/core/database/lib/entity-manager/__tests__/relations-sorter.test.js [hunk 6]
```diff
@@ -1,28 +1,44 @@
 'use strict';
 
-const relationsOrderer = require('../relations-sorter');
```
Comment (jhoward1994) [inline packages/core/database/lib/entity-manager/__tests__/relations-sorter.test.js:3]: Should we now rename this file to relations-orderer.test.js ?

File: packages/core/database/lib/entity-manager/relations-orderer.js (hunk 5 above)
Comment (Marc-Roig) [reply packages/core/database/lib/entity-manager/relations-orderer.js]: good catch! Done :)

**Review by jhoward1994 — APPROVED** [review]
LGTM 👍

File: packages/core/database/lib/entity-manager/index.js [hunk 7]
```diff
@@ -579,11 +579,23 @@ const createEntityManager = (db) => {
...
+            insert.forEach((data, idx) => {
+              data[orderColumnName] = idx + 1;
+            });
+          } else if (cleanRelationData.connect && hasOrderColumn(attribute)) {
+            // use position attributes to calculate order
+            const orderMap = relationsOrderer([], inverseJoinColumn.name, joinTable.orderColumnName)
+              .connect(relsToAdd)
+              .get()
+              // set the order based on the order of the ids
+              .reduce((acc, rel, idx) => ({ ...acc, [rel.id]: idx }), {});
```
Comment (petersg83) [inline packages/core/database/lib/entity-manager/index.js:592]: What is the difference with `getOrderMap();` ?

**Pull request merged**